File: src/web/magmac/app/stage/unit/UnitSet.ts
import { Iter } from "../../../../magmac/api/iter/Iter";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Function } from "../../../../java/util/function/Function";
export interface UnitSet<T> {
	 iterValues() : Iter<T>;
	 iter() : Iter<Unit<T>>;
	 add( element : Unit<T>) : UnitSet<T>;
	 mapValuesToResult( deserializer : Function<T, CompileResult<R>>) : CompileResult<UnitSet<R>>;
	 mapValues( serializer : Function<T, R>) : UnitSet<R>;
}

File: packages/compiler-java/src/main/java/magmac/app/stage/unit/UnitSet.java
package magmac.app.stage.unit;

import magmac.api.iter.Iter;
import magmac.app.compile.error.CompileResult;

import java.util.function.Function;

/**
 * Collection of compilation units that can be processed together.
 */
public interface UnitSet<T> {
    Iter<T> iterValues();

    Iter<Unit<T>> iter();

    UnitSet<T> add(Unit<T> element);

    <R> CompileResult<UnitSet<R>> mapValuesToResult(Function<T, CompileResult<R>> deserializer);

    <R> UnitSet<R> mapValues(Function<T, R> serializer);
}
